extensions/Bookalope/jsx/bookalope.jsx

```diff
@@ -193,3 +193,382 @@ function bookalopeCreateDocument(idmlFileName, bookId, bookflowId, betaHost) {
+            imageBounds[2] - frameBounds[0],
+            imageBounds[3] - frameBounds[1]
+        ];
+
+        // Then resize the image itself.
+        const newImageBoundY = anchor.geometricBounds[0] - (imageBounds[0] - imageBounds[0]);
+        const newImageBoundX = anchor.geometricBounds[1] - (imageBounds[1] - imageBounds[1]);
+        anchor.images[0].geometricBounds = [
+            newImageBoundY,
+            newImageBoundX,
+            newImageBoundY + (image.geometricBounds[2] - image.geometricBounds[0]),
+            newImageBoundX + (image.geometricBounds[3] - image.geometricBounds[1])
+        ];
+
+        // Adjust the textWrapPreferences for the anchored image.
+        anchor.textWrapPreferences.textWrapMode = imageRect.textWrapPreferences.textWrapMode;
+        anchor.textWrapPreferences.textWrapOffset = imageRect.textWrapPreferences.textWrapOffset;
+    }
+
+    /**
+     * Write the given Image object to disk using the given path and file name.
+     *
+     * @param {Graphic} graphic - The Graphic object to export.
+     * @param {File} file - The File object represents a local host file.
+     */
+    function exportPNG(graphic, file) {
+        try {
+            var other = graphic.duplicate();
+            try {
+                // FIXME Where's images documented?
+                // https://www.indesignjs.de/extendscriptAPI/indesign-latest/#Graphic.html
+                other.images[0].clearTransformations();
+                other.images[0].fit(FitOptions.FRAME_TO_CONTENT);
+            } catch (_) {
+                // Ignore any problem and try to export the image anyway.
+            }
+            other.exportFile(ExportFormat.PNG_FORMAT, file);
+            other.remove();
+        } catch (_) {
+            // Notify caller/user that exporting the image failed:
+            // https://github.com/bookalope/InDesign-CEP/pull/13#issuecomment-938668459
+        }
+    }
+
+    /**
+     * Compare the locations of two text frames based on their respective geometric
+     * boundaries (which have the format [y1, x1, y2, x2]). Returns -1 if frameA
+     * is above or left of frameB (i.e. is "smaller"); returns 1 otherwise.
+     *
+     * @param {TextFrame} frameA - A TextFrame object reference.
+     * @param {TextFrame} frameB - A TextFrame object reference.
+     * @return {Number} -1 if frameA is smaller than frameB, 1 otherwise.
+     */
+    function cmpFrames(frameA, frameB) {
+        if (frameA.geometricBounds[0] < frameB.geometricBounds[0]) {
+            return -1;
+        }
+        if (frameA.geometricBounds[0] == frameB.geometricBounds[0]) {
+            return (frameA.geometricBounds[1] < frameB.geometricBounds[1]) ? -1 : 1;
+        }
+        return 1;
+    }
+
+    // Create a temporary copy of the document that we want to export.
+    const tmpFile = File(app.createTemporaryCopy(doc.fullName))
+    const tmpDoc = app.open(tmpFile, false);
+    const tmpPath = tmpFile.path;
```

Comment:
On Windows, the [`app.createTemporaryCopy()`](https://www.indesignjs.de/extendscriptAPI/indesign-latest/#Application.html#d1e42253__d1e45703) creates a path _exactly like the original_, whereas on Mac it returns a path in the system’s temp folder. [I asked in Adobe’s Slack channel](https://adobedevs.slack.com/archives/C1FKLQ63F/p1635906342026700), let’s see what comes up. The result of that is that
```javascript
tmpDoc.close(SaveOptions.NO);
```
in [line 571](https://github.com/bookalope/InDesign-CEP/pull/13/files#diff-a7f0d70ffb3265087febcf37d8c4737fbac47f3bf7e9e1acfcc1d3f650407544R571) closes the active, open document. Sigh.

Reply:
Wow, that's quite a bug!

Reply:
Yes it’s weird and reproducible 😕 I wonder if we missed something, but it _should_ work the same way on both platforms?

Well, let’s see what the community responds in Slack.

Reply:
On my mac the script seems to work, but I get `Bookalope server error: Unable to set document because one is already set` from the UI

Reply:
Hm, that’s from calling [this endpoint](https://github.com/bookalope/Bookalope/blob/master/API.md#post-bookflows-files-document) here:

https://github.com/bookalope/InDesign-CEP/blob/4daddea57165a28163b6918d6bee743950ab7fd3/extensions/Bookalope/js/bookalope.js#L601-L607

What happens is that the extension uploads a document for a Bookflow that already has a document uploaded. Did you encounter an other error before that, and tried to upload again? Hmm 🤔

Reply:
> Well, let’s see what the community responds in Slack.

They asked me to file a bug, and suggested to work around it:

> That said, you should be able to pretty easily cook up your own “TempFileCopy()” (or whatever) command using standard ExtendScript code. These APIs would probably take you 99% of the way:
[File.copy()](https://github.com/bbb999/Types-for-Adobe/blob/c34df41331f7ff9cb4cd92352c8ecb2adbd449bf/shared/JavaScript.d.ts#L1651) (instance method)
[Folder.temp](https://github.com/bbb999/Types-for-Adobe/blob/c34df41331f7ff9cb4cd92352c8ecb2adbd449bf/shared/JavaScript.d.ts#L1854) (static property)